core/unix/os.c

```diff
@@ -6920,23 +6920,68 @@ pre_system_call(dcontext_t *dcontext)
          *     void __user *parent_tid, void __user *child_tid, struct pt_regs *regs)
          */
         uint64_t flags;
+        /* For the clone3 syscall, DR creates its own copy of clone_args for two
+         * reasons: to ensure that the app-provided clone_args is readable
+         * without any fault, and to avoid modifying the app's clone_args in the
+         * is_thread_create_syscall case (see below).
+         */
+        clone3_syscall_args_t *dr_clone_args = NULL, *app_clone_args = NULL;
+        uint app_clone_args_size = 0;
         if (dcontext->sys_num == SYS_clone3) {
-            clone3_syscall_args_t *cl_args = (clone3_syscall_args_t *)sys_param(
+            app_clone_args_size =
+                (uint)sys_param(dcontext, SYSCALL_PARAM_CLONE3_CLONE_ARGS_SIZE);
+            if (app_clone_args_size < CLONE_ARGS_SIZE_VER0) {
+                LOG(THREAD, LOG_SYSCALLS, 2, "\treturning EINVAL to app for clone3\n");
+                set_failure_return_val(dcontext, EINVAL);
+                DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+                execute_syscall = false;
+                break;
+            }
+            app_clone_args = (clone3_syscall_args_t *)sys_param(
                 dcontext, SYSCALL_PARAM_CLONE3_CLONE_ARGS);
-            flags = cl_args->flags;
-            /* Save for post_system_call. As clone3 flags are 64-bit, even in 32-bit
+            if (app_clone_args == NULL) {
+                LOG(THREAD, LOG_SYSCALLS, 2, "\treturning EFAULT to app for clone3\n");
+                set_failure_return_val(dcontext, EFAULT);
+                DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+                execute_syscall = false;
+                break;
+            }
+            /* Note that the struct clone_args being used by the app may have
+             * more fields than DR's internal struct (clone3_syscall_args_t).
+             * For creating DR's copy of the app's clone_args object, we need to
+             * allocate as much space as specified by the app in the clone3
+             * syscall's args.
+             */
+            dr_clone_args = (clone3_syscall_args_t *)heap_alloc(
+                dcontext, app_clone_args_size HEAPACCT(ACCT_OTHER));
+            if (!d_r_safe_read(app_clone_args, app_clone_args_size, dr_clone_args)) {
+                LOG(THREAD, LOG_SYSCALLS, 2, "\treturning EFAULT to app for clone3\n");
+                set_failure_return_val(dcontext, EFAULT);
+                DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+                execute_syscall = false;
+                heap_free(dcontext, dr_clone_args,
+                          app_clone_args_size HEAPACCT(ACCT_OTHER));
+                break;
+            }
+            flags = app_clone_args->flags;
```

Comment:
Done.

Curious about how this case works though. You mean the app might be intercepting faults to the page and after the first fault due to the safe read, it may mark it inaccessible?

Reply:
It may not be likely, and it may be difficult to write an app that hits the case (one example would be an app that has a thread that sits in a loop swapping the page protections between no-access and readable, and DR's safe read happens to occur when the page is readable) but it's a general TOCTOU error we should avoid to assume a page is accessible now just because it was a little while ago.  Besides, we have the copy, so there is no need to read the original.

Reply:
Where it was straight-forward, I changed the code to use the DR copy of clone args, instead of the app's. But there are some cases where it is not straight-forward. For `!CLONE_VM`, we do create a copy of the clone args to do the safe-read check (the check is required to read the `flags` field), but we free the copy pre-syscall. It is complex to keep it till post-syscall, as then we'd need to free it in the child and parent both.

There's also this check in `d_r_dispatch`; it seems like it might happen for the child too. We do not want to access the DR copy of clone args in the child thread, as there'll be a race with the parent free-ing it.
https://github.com/DynamoRIO/dynamorio/blob/eb2a333f859d5a9b89bf6d9dbb77f9f4b372811d/core/dispatch.c#L141

If the scenario you mentioned is unlikely, maybe it is better to avoid creating more complex code paths, and just live with multiple reads of the app clone args. Let me know what you think.

Reply:
I think we should avoid de-referencing app memory unsafely.  If we have to read it again, we should use another safe read.  Or we should store the info we need.  For was_thread_create_syscall, we just need the flags.  We could store that.  For !CLONE_VM I'm not sure what we need.

Reply:
> Or we should store the info we need. For was_thread_create_syscall, we just need the flags.

flags is all we need from the struct post-syscall, and storing just that would indeed be the simplest. Unfortunately the `sys_paramX` slots in `dcontext_t` are of type `reg_t`, but we need 8 bytes for storing the flags on 32-bit systems as well.

We could perhaps split the flags across two `sys_paramX` variables. This option will use up all of the 5 `sys_paramX` slots in total. 
sys_param0: app_clone_flags, to restore in the sys param reg post-syscall
sys_param1: dr_clone_flags, for freeing post-syscall
sys_param2: app_clone_flags_size, for freeing dr_clone_flags post-syscall
sys_param3, sys_param4: flags split across two vars.

> If we have to read it again, we should use another safe read

This is the other option. But we would need to pay for an extra safe read of the whole struct. I guess the previous option of splitting the flags is better.

Reply:
Made the change. PTAL.

If in future we need to save anything else from `clone_args` for post-syscall use, we may need to resort to the second option of using more safe-reads.

Reply:
Safe reads that do not fault are cheap so it is not a problem to do more of them.

Reply:
It's probably better to re-use the saved flags if possible, instead of doing a safe read each time and handling the possibility that the read may fault.

Reply:
So, for now I'll keep the current approach of using sys_param3 and sys_param4 for storing the 64-bit flags.